<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<link href="data:text/css;charset=utf-8,%0A%40font%2Dface%20%7B%0Afont%2Dfamily%3A%20octicons%2Dlink%3B%0Asrc%3A%20url%28data%3Afont%2Fwoff%3Bcharset%3Dutf%2D8%3Bbase64%2Cd09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM%2B8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB%2FaFGpk3jaTY6xa8JAGMW%2FO62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v%2Bk%2F0an2i%2BitHDw3v2%2B9%2BDBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3%2FI7AtxEJLtzzuZfI%2BVVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy%2FLt7Kc%2B0vWY%2FgAgIIEqAN9we0pwKXreiMasxvabDQMM4riO%2BqxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw%2Bymhce7vwM9jSqO8JyVd5RH9gyTt2%2BJ%2FyUmYlIR0s04n6%2B7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv%2FocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi%2BW2%2BMjCzMIDApSwvXzC97Z4Ig8N%2FBxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh%2F8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT%2BAEjAwuDFpBmA9KMDEwMCh9i%2Fv8H8sH0%2F4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9%2FlqYwOGZxeUelN2U2R6%2BcArgtCJpauW7UQBqnFkUsjAY%2FkOU1cP%2BDAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl%2BvvmM%2FbyA48e6tWrKArm4ZJlCbdsrxksL1AwWn%2FyBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO%2F%2FsdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd%2F89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF%2B9JOS0nbaaYDCQfwCJ7Au3AHj%2BLO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm%2BEBXuAbHmIMSRMs%2B4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL%2BhD7C1xoaHeLJSEao0FEW14ckxC%2BTU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13%2F%2Blm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl%2B9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O%2FAdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB%2F%2F%2FAA8AAQAAAAAAAAAAAAAAAAABAAAAAA%3D%3D%29%20format%28%27woff%27%29%3B%0A%7D%0Abody%20%7B%0A%2Dwebkit%2Dtext%2Dsize%2Dadjust%3A%20100%25%3B%0Atext%2Dsize%2Dadjust%3A%20100%25%3B%0Acolor%3A%20%23333%3B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20%22Segoe%20UI%22%2C%20Arial%2C%20freesans%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0Afont%2Dsize%3A%2016px%3B%0Aline%2Dheight%3A%201%2E6%3B%0Aword%2Dwrap%3A%20break%2Dword%3B%0A%7D%0Aa%20%7B%0Abackground%2Dcolor%3A%20transparent%3B%0A%7D%0Aa%3Aactive%2C%0Aa%3Ahover%20%7B%0Aoutline%3A%200%3B%0A%7D%0Astrong%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%202em%3B%0Amargin%3A%200%2E67em%200%3B%0A%7D%0Aimg%20%7B%0Aborder%3A%200%3B%0A%7D%0Ahr%20%7B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Aheight%3A%200%3B%0A%7D%0Apre%20%7B%0Aoverflow%3A%20auto%3B%0A%7D%0Acode%2C%0Akbd%2C%0Apre%20%7B%0Afont%2Dfamily%3A%20monospace%2C%20monospace%3B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ainput%20%7B%0Acolor%3A%20inherit%3B%0Afont%3A%20inherit%3B%0Amargin%3A%200%3B%0A%7D%0Ahtml%20input%5Bdisabled%5D%20%7B%0Acursor%3A%20default%3B%0A%7D%0Ainput%20%7B%0Aline%2Dheight%3A%20normal%3B%0A%7D%0Ainput%5Btype%3D%22checkbox%22%5D%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0Apadding%3A%200%3B%0A%7D%0Atable%20%7B%0Aborder%2Dcollapse%3A%20collapse%3B%0Aborder%2Dspacing%3A%200%3B%0A%7D%0Atd%2C%0Ath%20%7B%0Apadding%3A%200%3B%0A%7D%0A%2A%20%7B%0Abox%2Dsizing%3A%20border%2Dbox%3B%0A%7D%0Ainput%20%7B%0Afont%3A%2013px%20%2F%201%2E4%20Helvetica%2C%20arial%2C%20nimbussansl%2C%20liberationsans%2C%20freesans%2C%20clean%2C%20sans%2Dserif%2C%20%22Apple%20Color%20Emoji%22%2C%20%22Segoe%20UI%20Emoji%22%2C%20%22Segoe%20UI%20Symbol%22%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%234078c0%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%2C%0Aa%3Aactive%20%7B%0Atext%2Ddecoration%3A%20underline%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%200%3B%0Amargin%3A%2015px%200%3B%0Aoverflow%3A%20hidden%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23ddd%3B%0A%7D%0Ahr%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ahr%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%2015px%3B%0Amargin%2Dbottom%3A%2015px%3B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah1%20%7B%0Afont%2Dsize%3A%2030px%3B%0A%7D%0Ah2%20%7B%0Afont%2Dsize%3A%2021px%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%2016px%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%2011px%3B%0A%7D%0Ablockquote%20%7B%0Amargin%3A%200%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Aol%20ol%2C%0Aul%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Droman%3B%0A%7D%0Aul%20ul%20ol%2C%0Aul%20ol%20ol%2C%0Aol%20ul%20ol%2C%0Aol%20ol%20ol%20%7B%0Alist%2Dstyle%2Dtype%3A%20lower%2Dalpha%3B%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Afont%2Dsize%3A%2012px%3B%0A%7D%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0Afont%3A%2012px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0A%7D%0A%2Eselect%3A%3A%2Dms%2Dexpand%20%7B%0Aopacity%3A%200%3B%0A%7D%0A%2Eocticon%20%7B%0Afont%3A%20normal%20normal%20normal%2016px%2F1%20octicons%2Dlink%3B%0Adisplay%3A%20inline%2Dblock%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Drendering%3A%20auto%3B%0A%2Dwebkit%2Dfont%2Dsmoothing%3A%20antialiased%3B%0A%2Dmoz%2Dosx%2Dfont%2Dsmoothing%3A%20grayscale%3B%0A%2Dwebkit%2Duser%2Dselect%3A%20none%3B%0A%2Dmoz%2Duser%2Dselect%3A%20none%3B%0A%2Dms%2Duser%2Dselect%3A%20none%3B%0Auser%2Dselect%3A%20none%3B%0A%7D%0A%2Eocticon%2Dlink%3Abefore%20%7B%0Acontent%3A%20%27%5Cf05c%27%3B%0A%7D%0A%2Emarkdown%2Dbody%3Abefore%20%7B%0Adisplay%3A%20table%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3Aafter%20%7B%0Adisplay%3A%20table%3B%0Aclear%3A%20both%3B%0Acontent%3A%20%22%22%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%20%21important%3B%0A%7D%0A%2Emarkdown%2Dbody%3E%2A%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%20%21important%3B%0A%7D%0Aa%3Anot%28%5Bhref%5D%29%20%7B%0Acolor%3A%20inherit%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0A%2Eanchor%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%2Dright%3A%202px%3B%0Amargin%2Dleft%3A%20%2D18px%3B%0A%7D%0A%2Eanchor%3Afocus%20%7B%0Aoutline%3A%20none%3B%0A%7D%0Ah1%2C%0Ah2%2C%0Ah3%2C%0Ah4%2C%0Ah5%2C%0Ah6%20%7B%0Amargin%2Dtop%3A%201em%3B%0Amargin%2Dbottom%3A%2016px%3B%0Afont%2Dweight%3A%20bold%3B%0Aline%2Dheight%3A%201%2E4%3B%0A%7D%0Ah1%20%2Eocticon%2Dlink%2C%0Ah2%20%2Eocticon%2Dlink%2C%0Ah3%20%2Eocticon%2Dlink%2C%0Ah4%20%2Eocticon%2Dlink%2C%0Ah5%20%2Eocticon%2Dlink%2C%0Ah6%20%2Eocticon%2Dlink%20%7B%0Acolor%3A%20%23000%3B%0Avertical%2Dalign%3A%20middle%3B%0Avisibility%3A%20hidden%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%2C%0Ah2%3Ahover%20%2Eanchor%2C%0Ah3%3Ahover%20%2Eanchor%2C%0Ah4%3Ahover%20%2Eanchor%2C%0Ah5%3Ahover%20%2Eanchor%2C%0Ah6%3Ahover%20%2Eanchor%20%7B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Ah1%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah2%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah3%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah4%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah5%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%2C%0Ah6%3Ahover%20%2Eanchor%20%2Eocticon%2Dlink%20%7B%0Avisibility%3A%20visible%3B%0A%7D%0Ah1%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%202%2E25em%3B%0Aline%2Dheight%3A%201%2E2%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah1%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah2%20%7B%0Apadding%2Dbottom%3A%200%2E3em%3B%0Afont%2Dsize%3A%201%2E75em%3B%0Aline%2Dheight%3A%201%2E225%3B%0Aborder%2Dbottom%3A%201px%20solid%20%23eee%3B%0A%7D%0Ah2%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%3B%0A%7D%0Ah3%20%7B%0Afont%2Dsize%3A%201%2E5em%3B%0Aline%2Dheight%3A%201%2E43%3B%0A%7D%0Ah3%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah4%20%7B%0Afont%2Dsize%3A%201%2E25em%3B%0A%7D%0Ah4%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E2%3B%0A%7D%0Ah5%20%7B%0Afont%2Dsize%3A%201em%3B%0A%7D%0Ah5%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ah6%20%7B%0Afont%2Dsize%3A%201em%3B%0Acolor%3A%20%23777%3B%0A%7D%0Ah6%20%2Eanchor%20%7B%0Aline%2Dheight%3A%201%2E1%3B%0A%7D%0Ap%2C%0Ablockquote%2C%0Aul%2C%0Aol%2C%0Adl%2C%0Atable%2C%0Apre%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ahr%20%7B%0Aheight%3A%204px%3B%0Apadding%3A%200%3B%0Amargin%3A%2016px%200%3B%0Abackground%2Dcolor%3A%20%23e7e7e7%3B%0Aborder%3A%200%20none%3B%0A%7D%0Aul%2C%0Aol%20%7B%0Apadding%2Dleft%3A%202em%3B%0A%7D%0Aul%20ul%2C%0Aul%20ol%2C%0Aol%20ol%2C%0Aol%20ul%20%7B%0Amargin%2Dtop%3A%200%3B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Ali%3Ep%20%7B%0Amargin%2Dtop%3A%2016px%3B%0A%7D%0Adl%20%7B%0Apadding%3A%200%3B%0A%7D%0Adl%20dt%20%7B%0Apadding%3A%200%3B%0Amargin%2Dtop%3A%2016px%3B%0Afont%2Dsize%3A%201em%3B%0Afont%2Dstyle%3A%20italic%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Adl%20dd%20%7B%0Apadding%3A%200%2016px%3B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0Ablockquote%20%7B%0Apadding%3A%200%2015px%3B%0Acolor%3A%20%23777%3B%0Aborder%2Dleft%3A%204px%20solid%20%23ddd%3B%0A%7D%0Ablockquote%3E%3Afirst%2Dchild%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Ablockquote%3E%3Alast%2Dchild%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Atable%20%7B%0Adisplay%3A%20block%3B%0Awidth%3A%20100%25%3B%0Aoverflow%3A%20auto%3B%0Aword%2Dbreak%3A%20normal%3B%0Aword%2Dbreak%3A%20keep%2Dall%3B%0A%7D%0Atable%20th%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Atable%20th%2C%0Atable%20td%20%7B%0Apadding%3A%206px%2013px%3B%0Aborder%3A%201px%20solid%20%23ddd%3B%0A%7D%0Atable%20tr%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Aborder%2Dtop%3A%201px%20solid%20%23ccc%3B%0A%7D%0Atable%20tr%3Anth%2Dchild%282n%29%20%7B%0Abackground%2Dcolor%3A%20%23f8f8f8%3B%0A%7D%0Aimg%20%7B%0Amax%2Dwidth%3A%20100%25%3B%0Abox%2Dsizing%3A%20content%2Dbox%3B%0Abackground%2Dcolor%3A%20%23fff%3B%0A%7D%0Acode%20%7B%0Apadding%3A%200%3B%0Apadding%2Dtop%3A%200%2E2em%3B%0Apadding%2Dbottom%3A%200%2E2em%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%2085%25%3B%0Abackground%2Dcolor%3A%20rgba%280%2C0%2C0%2C0%2E04%29%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0Acode%3Abefore%2C%0Acode%3Aafter%20%7B%0Aletter%2Dspacing%3A%20%2D0%2E2em%3B%0Acontent%3A%20%22%5C00a0%22%3B%0A%7D%0Apre%3Ecode%20%7B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Afont%2Dsize%3A%20100%25%3B%0Aword%2Dbreak%3A%20normal%3B%0Awhite%2Dspace%3A%20pre%3B%0Abackground%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0A%2Ehighlight%20%7B%0Amargin%2Dbottom%3A%2016px%3B%0A%7D%0A%2Ehighlight%20pre%2C%0Apre%20%7B%0Apadding%3A%2016px%3B%0Aoverflow%3A%20auto%3B%0Afont%2Dsize%3A%2085%25%3B%0Aline%2Dheight%3A%201%2E45%3B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0A%7D%0A%2Ehighlight%20pre%20%7B%0Amargin%2Dbottom%3A%200%3B%0Aword%2Dbreak%3A%20normal%3B%0A%7D%0Apre%20%7B%0Aword%2Dwrap%3A%20normal%3B%0A%7D%0Apre%20code%20%7B%0Adisplay%3A%20inline%3B%0Amax%2Dwidth%3A%20initial%3B%0Apadding%3A%200%3B%0Amargin%3A%200%3B%0Aoverflow%3A%20initial%3B%0Aline%2Dheight%3A%20inherit%3B%0Aword%2Dwrap%3A%20normal%3B%0Abackground%2Dcolor%3A%20transparent%3B%0Aborder%3A%200%3B%0A%7D%0Apre%20code%3Abefore%2C%0Apre%20code%3Aafter%20%7B%0Acontent%3A%20normal%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%2Dsize%3A%2011px%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Epl%2Dc%20%7B%0Acolor%3A%20%23969896%3B%0A%7D%0A%2Epl%2Dc1%2C%0A%2Epl%2Ds%20%2Epl%2Dv%20%7B%0Acolor%3A%20%230086b3%3B%0A%7D%0A%2Epl%2De%2C%0A%2Epl%2Den%20%7B%0Acolor%3A%20%23795da3%3B%0A%7D%0A%2Epl%2Ds%20%2Epl%2Ds1%2C%0A%2Epl%2Dsmi%20%7B%0Acolor%3A%20%23333%3B%0A%7D%0A%2Epl%2Dent%20%7B%0Acolor%3A%20%2363a35c%3B%0A%7D%0A%2Epl%2Dk%20%7B%0Acolor%3A%20%23a71d5d%3B%0A%7D%0A%2Epl%2Dpds%2C%0A%2Epl%2Ds%2C%0A%2Epl%2Ds%20%2Epl%2Dpse%20%2Epl%2Ds1%2C%0A%2Epl%2Dsr%2C%0A%2Epl%2Dsr%20%2Epl%2Dcce%2C%0A%2Epl%2Dsr%20%2Epl%2Dsra%2C%0A%2Epl%2Dsr%20%2Epl%2Dsre%20%7B%0Acolor%3A%20%23183691%3B%0A%7D%0A%2Epl%2Dv%20%7B%0Acolor%3A%20%23ed6a43%3B%0A%7D%0A%2Epl%2Did%20%7B%0Acolor%3A%20%23b52a1d%3B%0A%7D%0A%2Epl%2Dii%20%7B%0Abackground%2Dcolor%3A%20%23b52a1d%3B%0Acolor%3A%20%23f8f8f8%3B%0A%7D%0A%2Epl%2Dsr%20%2Epl%2Dcce%20%7B%0Acolor%3A%20%2363a35c%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dml%20%7B%0Acolor%3A%20%23693a17%3B%0A%7D%0A%2Epl%2Dmh%2C%0A%2Epl%2Dmh%20%2Epl%2Den%2C%0A%2Epl%2Dms%20%7B%0Acolor%3A%20%231d3e81%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmq%20%7B%0Acolor%3A%20%23008080%3B%0A%7D%0A%2Epl%2Dmi%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dstyle%3A%20italic%3B%0A%7D%0A%2Epl%2Dmb%20%7B%0Acolor%3A%20%23333%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmd%20%7B%0Abackground%2Dcolor%3A%20%23ffecec%3B%0Acolor%3A%20%23bd2c00%3B%0A%7D%0A%2Epl%2Dmi1%20%7B%0Abackground%2Dcolor%3A%20%23eaffea%3B%0Acolor%3A%20%2355a532%3B%0A%7D%0A%2Epl%2Dmdr%20%7B%0Acolor%3A%20%23795da3%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%2Epl%2Dmo%20%7B%0Acolor%3A%20%231d3e81%3B%0A%7D%0Akbd%20%7B%0Adisplay%3A%20inline%2Dblock%3B%0Apadding%3A%203px%205px%3B%0Afont%3A%2011px%20Consolas%2C%20%22Liberation%20Mono%22%2C%20Menlo%2C%20Courier%2C%20monospace%3B%0Aline%2Dheight%3A%2010px%3B%0Acolor%3A%20%23555%3B%0Avertical%2Dalign%3A%20middle%3B%0Abackground%2Dcolor%3A%20%23fcfcfc%3B%0Aborder%3A%20solid%201px%20%23ccc%3B%0Aborder%2Dbottom%2Dcolor%3A%20%23bbb%3B%0Aborder%2Dradius%3A%203px%3B%0Abox%2Dshadow%3A%20inset%200%20%2D1px%200%20%23bbb%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20%7B%0Alist%2Dstyle%2Dtype%3A%20none%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%2B%2Etask%2Dlist%2Ditem%20%7B%0Amargin%2Dtop%3A%203px%3B%0A%7D%0A%2Etask%2Dlist%2Ditem%20input%20%7B%0Amargin%3A%200%200%2E35em%200%2E25em%20%2D1%2E6em%3B%0Avertical%2Dalign%3A%20middle%3B%0A%7D%0A%3Achecked%2B%2Eradio%2Dlabel%20%7B%0Az%2Dindex%3A%201%3B%0Aposition%3A%20relative%3B%0Aborder%2Dcolor%3A%20%234078c0%3B%0A%7D%0Acode%20%3E%20%2Ekw%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Edt%20%7B%20color%3A%20%23ed6a43%3B%20%7D%0Acode%20%3E%20%2Edv%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ebn%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Efl%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Ech%20%7B%20color%3A%20%23009999%3B%20%7D%0Acode%20%3E%20%2Est%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eco%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Eot%20%7B%20color%3A%20%230086b3%3B%20%7D%0Acode%20%3E%20%2Eal%20%7B%20color%3A%20%23a61717%3B%20%7D%0Acode%20%3E%20%2Efu%20%7B%20color%3A%20%2363a35c%3B%20%7D%0Acode%20%3E%20%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%0Acode%20%3E%20%2Ewa%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ecn%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Esc%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Evs%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Ess%20%7B%20color%3A%20%23183691%3B%20%7D%0Acode%20%3E%20%2Eim%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eva%20%7Bcolor%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecf%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eop%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Ebu%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Eex%20%7B%20color%3A%20%23000000%3B%20%7D%0Acode%20%3E%20%2Epp%20%7B%20color%3A%20%23999999%3B%20%7D%0Acode%20%3E%20%2Eat%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Edo%20%7B%20color%3A%20%23969896%3B%20%7D%0Acode%20%3E%20%2Ean%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ecv%20%7B%20color%3A%20%23008080%3B%20%7D%0Acode%20%3E%20%2Ein%20%7B%20color%3A%20%23008080%3B%20%7D%0A" rel="stylesheet">
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="readme">README</h1>
<p>Maria Izabel cavassim Alves<br />
10/26/2017</p>
<ul>
<li><a href="#introduction">Introduction</a>
<ul>
<li><a href="#genome-assembly-spades">Genome assembly: Spades</a></li>
<li><a href="#assembly-stats-quast">Assembly stats: QUAST</a></li>
<li><a href="#gene-identification-prokka">Gene identification: Prokka</a></li>
<li><a href="#gene-alignment-clustalo">Gene alignment: ClustalO</a></li>
<li><a href="#snp-calling">SNP calling</a></li>
<li><a href="#orthologous-identification-proteinortho">Orthologous identification: ProteinOrtho</a></li>
<li><a href="#orthologous-identification-orthomcl">Orthologous identification: OrthoMCL</a></li>
</ul></li>
</ul>
<h1 id="introduction">Introduction</h1>
<p>This document was created as a resource of the scripts used in the analysis of Rhizobium paper.</p>
<h2 id="genome-assembly-spades">Genome assembly: Spades</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="kw">for</span> <span class="kw">f</span> in ../reads/*_1_trimmed.fastq.gz<span class="kw">;</span> <span class="kw">do</span>
    <span class="ot">FILE=${f#</span>../reads/*<span class="ot">}</span>
    <span class="ot">STRAIN=${FILE%</span>*_1_trimmed.fastq.gz<span class="ot">}</span>
    <span class="kw">if [</span> <span class="ot">-d</span> <span class="ot">$STRAIN</span><span class="st">&quot;/&quot;</span><span class="ot">$STRAIN</span><span class="st">&quot;_spades_careful&quot;</span><span class="kw"> ]</span>; <span class="kw">then</span>
        <span class="kw">echo</span> <span class="ot">$STRAIN</span><span class="st">&quot;: _spades_careful directory exists. Skipping strain.&quot;</span>
        <span class="kw">continue</span>
    <span class="kw">fi</span>

    <span class="kw">mkdir</span> -p <span class="ot">$STRAIN</span>
    <span class="kw">qx</span> --no-scratch -c 8 -m 32g -t 05:00:00 <span class="st">&quot;/project/clover/faststorage/tools/SPAdes-3.6.2-Linux/bin/spades.py --careful -t 8 -m 32 -o </span><span class="ot">$STRAIN</span><span class="st">&quot;</span>/<span class="st">&quot;</span><span class="ot">$STRAIN</span><span class="st">&quot;</span>_spades_careful<span class="st">&quot; -1 </span><span class="ot">$f</span><span class="st"> -2 </span><span class="ot">${f%</span>*1_trimmed.fastq.gz<span class="ot">}</span><span class="st">&quot;</span>2_trimmed.fastq.gz<span class="st">&quot; -s </span><span class="ot">${f%</span>*1_trimmed.fastq.gz<span class="ot">}</span><span class="st">&quot;</span>U1_trimmed.fastq.gz<span class="st">&quot; -s </span><span class="ot">${f%</span>*1_trimmed.fastq.gz<span class="ot">}</span><span class="st">&quot;</span>U2_trimmed.fastq.gz<span class="st">&quot;&quot;</span><span class="kw">;</span>
<span class="kw">done</span></code></pre></div>
<h2 id="assembly-stats-quast">Assembly stats: QUAST</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="co"># For each strain, create a quast report over SPAdes, a5, Discovar and CISA assemblies.</span>
<span class="kw">for</span> <span class="kw">f</span> in contigs/*.fasta<span class="kw">;</span> <span class="kw">do</span>
    <span class="ot">FILE_NAME=${f#</span>*/<span class="ot">}</span>
    <span class="ot">STRAIN_NAME=${FILE_NAME%</span>*.fasta<span class="ot">}</span>
    <span class="kw">qx</span> --no-scratch -c 1 -t 00:10:00 <span class="st">&quot;../../tools/quast-3.2/quast.py -t 1 -o assemblies/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;/quast contigs/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;.fasta assemblies/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;/spades/contigs.fasta assemblies/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;/a5/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;.contigs.fasta assemblies/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;/discovar/a.final/a.lines.fasta assemblies/&quot;</span><span class="ot">$STRAIN_NAME</span><span class="st">&quot;/cisa/CISA_Out.fa&quot;</span>
<span class="kw">done</span></code></pre></div>
<h2 id="gene-identification-prokka">Gene identification: Prokka</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="co"># For each contig run prokka, many output files are generated: .gff, .gbk, .fna, .faa and so on.</span>
<span class="kw">for</span> <span class="kw">f</span> in */*.contigs.fna<span class="kw">;</span> <span class="kw">do</span>
    <span class="ot">STRAIN=$(IFS=</span><span class="st">&quot;/&quot;</span>; <span class="kw">set</span> <span class="kw">--</span> <span class="ot">$f</span><span class="kw">;</span> <span class="kw">echo</span> <span class="ot">$1)</span>
    <span class="kw">if [</span> <span class="ot">-f</span> <span class="ot">$STRAIN</span><span class="st">&quot;/&quot;</span><span class="ot">$STRAIN</span><span class="st">&quot;.gff&quot;</span><span class="kw"> ]</span>; <span class="kw">then</span>
    <span class="kw">echo</span> <span class="ot">$STRAIN</span><span class="st">&quot;: gff file exists. Skipping strain.&quot;</span>
    <span class="kw">continue</span>
    <span class="kw">fi</span>
    <span class="kw">cd</span> <span class="ot">$STRAIN</span>
    <span class="kw">qx</span> --no-scratch -t 10:00:00 -c 4 <span class="st">&quot;/project/clover/faststorage/tools/prokka-1.11/bin/prokka -cpus=4 *.contigs.fna&quot;</span>
    <span class="kw">cd</span> ..
<span class="kw">done</span></code></pre></div>
<h2 id="gene-alignment-clustalo">Gene alignment: ClustalO</h2>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">
<span class="co"># Group fasta files are aligned in a ???codon aware matter???. The clustalO runs with default parameters. </span>

<span class="im">import</span> argparse
<span class="im">import</span> io
<span class="im">import</span> subprocess
<span class="im">from</span> itertools <span class="im">import</span> groupby

<span class="im">from</span> Bio <span class="im">import</span> AlignIO
<span class="im">from</span> Bio.Align <span class="im">import</span> MultipleSeqAlignment
<span class="im">from</span> Bio.Alphabet <span class="im">import</span> generic_dna
<span class="im">from</span> Bio.Seq <span class="im">import</span> Seq
<span class="im">from</span> Bio.SeqRecord <span class="im">import</span> SeqRecord


<span class="kw">def</span> fasta_iter(fasta_file):
    <span class="co">&quot;&quot;&quot;</span>
<span class="co">    Given a FASTA file, yield tuples of header, sequence.</span>
<span class="co">    Author: Brent Pedersen (https://www.biostars.org/p/710/#1412)</span>
<span class="co">    &quot;&quot;&quot;</span>
    fh <span class="op">=</span> <span class="bu">open</span>(fasta_file)
    <span class="co"># Ditch the boolean (x[0]) and just keep the header or sequence since we know they alternate.</span>
    faiter <span class="op">=</span> (x[<span class="dv">1</span>] <span class="cf">for</span> x <span class="op">in</span> groupby(fh, <span class="kw">lambda</span> line: line[<span class="dv">0</span>] <span class="op">==</span> <span class="st">&quot;&gt;&quot;</span>))
    <span class="cf">for</span> header <span class="op">in</span> faiter:
        <span class="co"># Drop the &quot;&gt;&quot;.</span>
        header <span class="op">=</span> header.<span class="fu">__next__</span>()[<span class="dv">1</span>:].strip()
        <span class="co"># Join all sequence lines to one in case it spans multiple lines.</span>
        seq <span class="op">=</span> <span class="st">&quot;&quot;</span>.join(s.strip() <span class="cf">for</span> s <span class="op">in</span> faiter.<span class="fu">__next__</span>())
        <span class="cf">yield</span> header, seq

<span class="kw">def</span> seq_to_str(header, seq):
    <span class="cf">return</span> <span class="st">&quot;&gt;{}</span><span class="ch">\n</span><span class="st">{}</span><span class="ch">\n</span><span class="st">&quot;</span>.<span class="bu">format</span>(header, seq)

<span class="kw">def</span> back_translate(fasta_aln, original_sequences):
    string_handle <span class="op">=</span> io.StringIO()
    string_handle.write(fasta_aln)
    string_handle.seek(<span class="dv">0</span>)
    outer_alignments <span class="op">=</span> AlignIO.parse(string_handle, <span class="st">&quot;fasta&quot;</span>)
    back_translated_records <span class="op">=</span> []
    <span class="cf">for</span> alignments <span class="op">in</span> outer_alignments:
        <span class="cf">for</span> aln <span class="op">in</span> alignments:
            original_seq <span class="op">=</span> original_sequences[aln.<span class="bu">id</span>]

            back_translated <span class="op">=</span> []
            gaps <span class="op">=</span> <span class="dv">0</span>
            <span class="cf">for</span> i <span class="op">in</span> <span class="bu">range</span>(<span class="bu">len</span>(aln.seq)):
                <span class="cf">if</span> aln.seq[i] <span class="op">==</span> <span class="st">&quot;-&quot;</span>:
                    gaps <span class="op">+=</span> <span class="dv">1</span>
                    back_translated.append(<span class="st">&quot;---&quot;</span>)
                <span class="cf">else</span>:
                    idx <span class="op">=</span> (i <span class="op">-</span> gaps) <span class="op">*</span> <span class="dv">3</span>
                    back_translated.append(original_seq[idx:(idx <span class="op">+</span> <span class="dv">3</span>)])

            back_translated_records.append(SeqRecord(Seq(<span class="st">&quot;&quot;</span>.join(back_translated), generic_dna), aln.<span class="bu">id</span>))

    <span class="cf">return</span> MultipleSeqAlignment(back_translated_records, alphabet<span class="op">=</span>generic_dna)


<span class="kw">def</span> msa_to_str(msa):
    <span class="co"># msa.format(&quot;fasta&quot;) prints unecessary things such as description. This is a simpler version.</span>
    <span class="cf">return</span> <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>.join(seq_to_str(aln.<span class="bu">id</span>, aln.seq) <span class="cf">for</span> aln <span class="op">in</span> <span class="bu">sorted</span>(msa, key<span class="op">=</span><span class="kw">lambda</span> aln: aln.<span class="bu">id</span>))


<span class="kw">def</span> main():
    parser <span class="op">=</span> argparse.ArgumentParser(description<span class="op">=</span><span class="st">&quot;Read a fasta file and align the sequences in it using ClustalO in a &quot;</span>
                                                 <span class="co">&quot;codon-aware style.&quot;</span>)
    parser.add_argument(<span class="st">&quot;in_file&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the input fasta file&quot;</span>)
    parser.add_argument(<span class="st">&quot;clustal_o_exec&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the path to the ClustalO executeable&quot;</span>, default<span class="op">=</span><span class="st">&quot;clustalo&quot;</span>)
    parser.add_argument(<span class="st">'--output_aa'</span>, dest<span class="op">=</span><span class="st">'output_aa'</span>, action<span class="op">=</span><span class="st">'store_true'</span>)
    parser.add_argument(<span class="st">&quot;--threads&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">int</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the number of threads to use&quot;</span>, default<span class="op">=</span><span class="dv">1</span>)
    args <span class="op">=</span> parser.parse_args()

    original_sequences <span class="op">=</span> {}
    <span class="cf">for</span> h, s <span class="op">in</span> fasta_iter(args.in_file):
        orig_h <span class="op">=</span> h
        i <span class="op">=</span> <span class="dv">2</span>
        <span class="cf">while</span> h <span class="op">in</span> original_sequences:
            h <span class="op">=</span> <span class="st">&quot;{}_{}&quot;</span>.<span class="bu">format</span>(orig_h, i)
            i <span class="op">+=</span> <span class="dv">1</span>
        original_sequences[h] <span class="op">=</span> s

    translated_sequences <span class="op">=</span> [(header, Seq(seq).translate()) <span class="cf">for</span> header, seq <span class="op">in</span> original_sequences.items()]
    p <span class="op">=</span> subprocess.Popen([args.clustal_o_exec, <span class="st">&quot;--outfmt=fasta&quot;</span>, <span class="st">&quot;--seqtype=Protein&quot;</span>,
                          <span class="co">&quot;--threads={}&quot;</span>.<span class="bu">format</span>(<span class="bu">str</span>(args.threads)), <span class="st">&quot;--infile=-&quot;</span>],
                         stdout<span class="op">=</span>subprocess.PIPE, stdin<span class="op">=</span>subprocess.PIPE, stderr<span class="op">=</span>subprocess.STDOUT)
    aln_out <span class="op">=</span> p.communicate(<span class="bu">input</span><span class="op">=</span><span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>.join(seq_to_str(h, s) <span class="cf">for</span> h, s <span class="op">in</span> translated_sequences).encode())
    <span class="cf">if</span> args.output_aa:
        <span class="bu">print</span>(aln_out[<span class="dv">0</span>].decode())
    <span class="cf">else</span>:
        back_translated_alignment <span class="op">=</span> back_translate(aln_out[<span class="dv">0</span>].decode(), original_sequences)
        <span class="bu">print</span>(msa_to_str(back_translated_alignment))


<span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">'__main__'</span>:
    main()</code></pre></div>
<h2 id="snp-calling">SNP calling</h2>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="co"># Script located in scripts_methods ld.py</span></code></pre></div>
<h2 id="orthologous-identification-proteinortho">Orthologous identification: ProteinOrtho</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="co"># For each protein fasta file, go through each step of proteinortho, with synteny flag enabled</span>

<span class="kw">qx</span> --no-scratch -t 00:30:00 -c 16 -w <span class="st">&quot;/home/agb/tools/proteinortho_v5.13/proteinortho5.pl -cpus=16 -synteny -step=1 data/*.faa&quot;</span>

<span class="kw">for</span> <span class="kw">i</span> in <span class="dt">{1..199}</span><span class="kw">;</span> <span class="kw">do</span>
    <span class="kw">qx</span> --no-scratch -t 24:00:00 -c 4 <span class="st">&quot;/home/agb/tools/proteinortho_v5.13/proteinortho5.pl -cpus=4 -synteny -step=2 -startat=</span><span class="ot">$i</span><span class="st"> -stopat=</span><span class="ot">$i</span><span class="st"> data/*.faa&quot;</span>
<span class="kw">done</span>

<span class="kw">qx</span> --no-scratch -t 4:00:00 -c 8 <span class="st">&quot;/home/agb/tools/proteinortho_v5.13/proteinortho5.pl -cpus=8 -synteny -step=3 data/*.faa&quot;</span></code></pre></div>
<h2 id="orthologous-identification-orthomcl">Orthologous identification: OrthoMCL</h2>
<p>OrthoMCL analysis were executed with PosgresSQL instead of mySQL. More details on how to run OrthoMCL using PostgreSQL can be found [here]<a href="https://bitbucket.org/asger/orthomcl-postgresql" class="uri">https://bitbucket.org/asger/orthomcl-postgresql</a>. The output of OrthoMCL it was then parsed into fasta files. The script ??disambiguate_orthomcl_groups.py?? is then used to filter for paralagous genes.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python">
<span class="im">import</span> argparse
<span class="im">import</span> glob
<span class="im">from</span> itertools <span class="im">import</span> groupby

<span class="im">from</span> GFFReader <span class="im">import</span> gff_iter
<span class="im">from</span> disambiguate_orthomcl_groups <span class="im">import</span> get_disambiguated_groups, Gene


<span class="kw">def</span> fasta_iter(fasta_file):
    <span class="co">&quot;&quot;&quot;</span>
<span class="co">    Given a FASTA file, yield tuples of header, sequence.</span>
<span class="co">    Author: Brent Pedersen (https://www.biostars.org/p/710/#1412)</span>
<span class="co">    &quot;&quot;&quot;</span>
    fh <span class="op">=</span> <span class="bu">open</span>(fasta_file)
    <span class="co"># Ditch the boolean (x[0]) and just keep the header or sequence since we know they alternate.</span>
    faiter <span class="op">=</span> (x[<span class="dv">1</span>] <span class="cf">for</span> x <span class="op">in</span> groupby(fh, <span class="kw">lambda</span> line: line[<span class="dv">0</span>] <span class="op">==</span> <span class="st">&quot;&gt;&quot;</span>))
    <span class="cf">for</span> header <span class="op">in</span> faiter:
        <span class="co"># Drop the &quot;&gt;&quot;.</span>
        header <span class="op">=</span> header.<span class="fu">__next__</span>()[<span class="dv">1</span>:].strip()
        <span class="co"># Join all sequence lines to one in case it spans multiple lines.</span>
        seq <span class="op">=</span> <span class="st">&quot;&quot;</span>.join(s.strip() <span class="cf">for</span> s <span class="op">in</span> faiter.<span class="fu">__next__</span>())
        <span class="cf">yield</span> header, seq


<span class="kw">def</span> reverse_complement(seq):
    seq_dict <span class="op">=</span> {<span class="st">'A'</span>: <span class="st">'T'</span>, <span class="st">'T'</span>: <span class="st">'A'</span>, <span class="st">'G'</span>: <span class="st">'C'</span>, <span class="st">'C'</span>: <span class="st">'G'</span>}
    <span class="cf">return</span> <span class="st">&quot;&quot;</span>.join([seq_dict[base] <span class="cf">for</span> base <span class="op">in</span> <span class="bu">reversed</span>(seq)])


<span class="kw">def</span> get_genes_from_gff(gff_glob):
    genes <span class="op">=</span> {}
    <span class="cf">for</span> gff_file_name <span class="op">in</span> glob.iglob(gff_glob):
        gff_glob_file_name <span class="op">=</span> gff_file_name[gff_file_name.rfind(<span class="st">&quot;/&quot;</span>) <span class="op">+</span> <span class="dv">1</span>:]
        strain_name <span class="op">=</span> gff_glob_file_name[:gff_glob_file_name.find(<span class="st">&quot;.&quot;</span>)][<span class="dv">0</span>:<span class="dv">4</span>]
        <span class="bu">print</span>(<span class="st">&quot;Reading genes for strain {} (file {})...&quot;</span>.<span class="bu">format</span>(strain_name, gff_file_name))

        genes[strain_name] <span class="op">=</span> {}

        previous_gene <span class="op">=</span> <span class="va">None</span>
        <span class="cf">for</span> gff_entry <span class="op">in</span> gff_iter(gff_file_name, type_filter<span class="op">=</span><span class="st">&quot;CDS&quot;</span>):
            gene_id <span class="op">=</span> gff_entry.attributes[<span class="st">&quot;ID&quot;</span>].replace(<span class="st">&quot;PROKKA_&quot;</span>, <span class="st">&quot;&quot;</span>)
            gene <span class="op">=</span> Gene(gene_id, gff_entry.sequence, gff_entry.start, gff_entry.end, gff_entry.strand <span class="op">==</span> <span class="st">&quot;-&quot;</span>)
            <span class="cf">if</span> previous_gene <span class="op">is</span> <span class="op">not</span> <span class="va">None</span> <span class="op">and</span> previous_gene.seq_name <span class="op">==</span> gff_entry.sequence:
                gene.neighbor_l <span class="op">=</span> previous_gene
                previous_gene.neighbor_r <span class="op">=</span> gene

            genes[strain_name][gff_entry.attributes[<span class="st">&quot;ID&quot;</span>]] <span class="op">=</span> gene
            previous_gene <span class="op">=</span> gene

    <span class="cf">return</span> genes


<span class="kw">def</span> read_orthomcl_groups(file_name, ortho_name_to_strain_name<span class="op">=</span><span class="va">None</span>):
    groups <span class="op">=</span> []
    <span class="cf">with</span> <span class="bu">open</span>(file_name) <span class="im">as</span> file_h:
        <span class="cf">for</span> line <span class="op">in</span> file_h:
            <span class="cf">if</span> line.isspace():
                <span class="cf">continue</span>

            line_members <span class="op">=</span> line.strip().split(<span class="st">&quot; &quot;</span>)
            <span class="cf">if</span> ortho_name_to_strain_name <span class="op">is</span> <span class="op">not</span> <span class="va">None</span>:
                group_members <span class="op">=</span> [(ortho_name_to_strain_name[m.split(<span class="st">&quot;|&quot;</span>)[<span class="dv">0</span>]], m.split(<span class="st">&quot;|&quot;</span>)[<span class="dv">1</span>])
                                 <span class="cf">for</span> m <span class="op">in</span> line_members[<span class="dv">1</span>:]]
            <span class="cf">else</span>:
                group_members <span class="op">=</span> [(m.split(<span class="st">&quot;|&quot;</span>)[<span class="dv">0</span>], m.split(<span class="st">&quot;|&quot;</span>)[<span class="dv">1</span>]) <span class="cf">for</span> m <span class="op">in</span> line_members[<span class="dv">1</span>:]]
            group_id <span class="op">=</span> line_members[<span class="dv">0</span>][:<span class="op">-</span><span class="dv">1</span>]
            groups.append((group_id, group_members))

    <span class="cf">return</span> groups


<span class="kw">def</span> main():
    parser <span class="op">=</span> argparse.ArgumentParser(description<span class="op">=</span><span class="st">&quot;Read an OrthoMCL output file and output each group as a fasta file.&quot;</span>)
    parser.add_argument(<span class="st">&quot;in_file&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the input (groups.txt) file&quot;</span>)
    parser.add_argument(<span class="st">&quot;fna_glob&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;a glob pattern pointing to all fna files to load.&quot;</span>
                                                   <span class="co">&quot; Remember to quote your argument to stop wildcard expansion.&quot;</span>)
    parser.add_argument(<span class="st">&quot;gff_glob&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;a glob pattern pointing to all the gff files to load.&quot;</span>
                                                   <span class="co">&quot; Remember to quote your argument to stop wildcard expansion.&quot;</span>)
    parser.add_argument(<span class="st">&quot;u_out_files_dir&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the output directory of the unambiguous group fasta files&quot;</span>)
    parser.add_argument(<span class="st">&quot;a_out_files_dir&quot;</span>, <span class="bu">type</span><span class="op">=</span><span class="bu">str</span>, <span class="bu">help</span><span class="op">=</span><span class="st">&quot;the output directory of the ambiguous group fasta files&quot;</span>)

    args <span class="op">=</span> parser.parse_args()
    <span class="cf">assert</span> args.fna_glob.startswith(<span class="st">&quot;~&quot;</span>), <span class="st">&quot;You cannot use '~' in the fna glob pattern.&quot;</span>
    <span class="cf">assert</span> args.gff_glob.startswith(<span class="st">&quot;~&quot;</span>), <span class="st">&quot;You cannot use '~' in the gff glob pattern.&quot;</span>

    <span class="co"># Extract the sequences from each strain's fasta file. Assume we can keep it all in memory!</span>
    sequences <span class="op">=</span> {}
    <span class="bu">print</span>(<span class="st">&quot;Using fna glob '{}'.&quot;</span>.<span class="bu">format</span>(args.fna_glob))
    <span class="cf">for</span> strain_file_name <span class="op">in</span> glob.iglob(args.fna_glob):
        fna_glob_file_name <span class="op">=</span> strain_file_name[strain_file_name.rfind(<span class="st">&quot;/&quot;</span>) <span class="op">+</span> <span class="dv">1</span>:]
        strain_name <span class="op">=</span> fna_glob_file_name[:fna_glob_file_name.find(<span class="st">&quot;.&quot;</span>)][<span class="dv">0</span>:<span class="dv">4</span>]
        <span class="bu">print</span>(<span class="st">&quot;Reading sequences for strain {} (file {})...&quot;</span>.<span class="bu">format</span>(strain_name, strain_file_name))
        sequences[strain_name] <span class="op">=</span> {}
        <span class="cf">for</span> seq_record <span class="op">in</span> fasta_iter(strain_file_name):
            seq_id <span class="op">=</span> seq_record[<span class="dv">0</span>].split(<span class="st">&quot; &quot;</span>)[<span class="dv">0</span>]
            sequences[strain_name][seq_id] <span class="op">=</span> seq_record[<span class="dv">1</span>]

    <span class="bu">print</span>(<span class="st">&quot;Using gff glob '{}'.&quot;</span>.<span class="bu">format</span>(args.gff_glob))
    genes <span class="op">=</span> get_genes_from_gff(args.gff_glob)

    <span class="cf">assert</span> <span class="bu">set</span>(sequences.keys()) <span class="op">==</span> <span class="bu">set</span>(genes.keys())

    groups <span class="op">=</span> read_orthomcl_groups(args.in_file)
    groups, ambiguous_groups, unambiguous_groups <span class="op">=</span> get_disambiguated_groups(groups, genes, verbose<span class="op">=</span><span class="va">True</span>)

    one_based <span class="op">=</span> <span class="va">True</span>
    start_pos_offset <span class="op">=</span> <span class="dv">1</span> <span class="cf">if</span> one_based <span class="cf">else</span> <span class="dv">0</span>

    <span class="cf">for</span> group_id, group_members <span class="op">in</span> unambiguous_groups:
        write_group_fasta_file(args.u_out_files_dir, genes, group_id, group_members, sequences, start_pos_offset)
    <span class="cf">for</span> group_id, group_members <span class="op">in</span> ambiguous_groups:
        write_group_fasta_file(args.a_out_files_dir, genes, group_id, group_members, sequences, start_pos_offset)


<span class="kw">def</span> write_group_fasta_file(out_files_dir, genes, group_id, group_members, sequences, start_pos_offset):
    fasta_strs <span class="op">=</span> []
    <span class="cf">for</span> strain, gs <span class="op">in</span> group_members.items():
        <span class="cf">for</span> gene_name <span class="op">in</span> gs:
            gene <span class="op">=</span> genes[strain][gene_name]
            start, end <span class="op">=</span> <span class="bu">min</span>(gene.start, gene.end), <span class="bu">max</span>(gene.start, gene.end)
            gene_subseq <span class="op">=</span> sequences[strain][gene.seq_name][(start <span class="op">-</span> start_pos_offset):end]
            <span class="cf">if</span> gene.reverse:
                gene_subseq <span class="op">=</span> reverse_complement(gene_subseq)

            header <span class="op">=</span> <span class="st">&quot;{}|{}|{}&quot;</span>.<span class="bu">format</span>(group_id, strain, gene.name)
            fasta_strs.append(<span class="st">&quot;&gt;{}</span><span class="ch">\n</span><span class="st">{}</span><span class="ch">\n</span><span class="st">&quot;</span>.<span class="bu">format</span>(header, gene_subseq))
    <span class="cf">with</span> <span class="bu">open</span>(<span class="st">&quot;{}/{}.fna&quot;</span>.<span class="bu">format</span>(out_files_dir, group_id), <span class="st">&quot;w&quot;</span>) <span class="im">as</span> out_file:
        out_file.write(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>.join(fasta_strs))


<span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">'__main__'</span>:
    main()

<span class="co"># for f in group_fasta_files/u/group1000.fna; do</span>
<span class="co">#     OUT=&quot;group_fasta_files/u/alns/&quot;${f#group_fasta_files/u/*}</span>
<span class="co">#     echo &quot;$f - $OUT&quot;</span>
<span class="co">#     qx --no-scratch -t 00:15:00 &quot;/home/agb/tools/prank -d=$f -o=$OUT -codon&quot;</span>
<span class="co"># done</span></code></pre></div>

</body>
</html>
